### PR TITLE
Reflective particle boundary condition

### PR DESCRIPTION
**Pull request by dpgrote**
This PR does two related things. It implements a reflective particle boundary condition, where particles that leave the grid are put back in and have the sign of the their velocity changed. The second thing is to allow differing boundary conditions at each side of the domain, for example reflecting on one side, absorbing on the other. This ended up being more complicated than I at first thought, needing the added file ParticleBoundaries.H.

It adds the option 'Reflecting' to the boundary.particle_lo and _hi input parameters.

Is a CI test needed for this? I can construct a simple test checking a single particle's motion.

A notable change is making `ParticleBoundaryType` an `enum struct`. This seems to be more robust since it will do compile time type checking of the enum (instead of using plain ints). If there is a reason for not doing this, I'll switch it back to the previous struct {enum {}}.

The flag `boundary.reflect_all_velocities` is added which determines whether the sign of only the normal velocity or all velocities is changed upon reflection. Setting this to true gives cleaner behavior when there are magnetic fields at the boundary.

### CODE REVIEW

**Comment by RemiLehe** [comment]
@dpgrote Sorry for the very late review! It seems that there is a conflict ; could you update this PR?

**Comment by dpgrote** [comment]
Hi @RemiLehe Thanks, the conflict is fixed.

**Comment by ax3l** [comment]
I kicked off the Python CI test again in case that was a temporary issue.
There is also a compile problem with `Python_Langmuir_rz_multimode` it seems.

> Is a CI test needed for this? I can construct a simple test checking a single particle's motion.

That would be splendid :+1:

**Comment by dpgrote** [comment]
@ax3l @RemiLehe Any ideas why the Azure CI test is failing? The compile errors it is showing don't make sense.

**Comment by dpgrote** [comment]
> @ax3l @RemiLehe Any ideas why the Azure CI test is failing? The compile errors it is showing don't make sense.

Nevermind, I was making a dumb mistake.

File: Docs/source/usage/parameters.rst [hunk 1]
```diff
@@ -387,9 +392,17 @@ Particle initialization
     Not currently available in the RZ configuration.
 
 * ``particles.boundary_conditions`` (`string`) optional (default `none`)
-    Boundary conditions applied to particles. Options are:
+    Boundary conditions to apply to particles on all boundaries. Options are:
     * ``none``: the boundary conditions applied to particles is determined by ``geometry.is_periodic``.
     * ``absorbing``: particles exiting the simulation domain are discarded.
+    * ``reflecting``: particles are reflected from the boundaries
```
Comment (RemiLehe) [inline Docs/source/usage/parameters.rst]: I am a bit confused: does `particles.boundary_conditions` duplicates the parameter `boundary.particle_lo`?
If yes, will `particle.boundary_conditions` be deprecated at some point?

File: Examples/Tests/boundaries/analysis.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+# Copyright 2021 David Grote
+#
+# This file is part of WarpX.
+#
+# License: BSD-3-Clause-LBNL
+
+
+"""
+This script tests the particle boundary conditions
```
Comment (RemiLehe) [inline Examples/Tests/boundaries/analysis.py]: Could you add an overall description of what the test does here?
(e.g. summary of the overall physical setup, and what gets checked at the end)

File: Examples/Tests/boundaries/analysis.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+xxa[0] = do_reflect(xxa[0])
+xxa[1] = do_reflect(xxa[1])
+
+zza = zz0 + vz0*time
+zza[0] = do_periodic(zza[0])
+zza[1] = do_periodic(zza[1])
+
+assert (len(a_id) == 1), 'Absorbing particles not absorbed'
+assert (np.all(vx == -vx0)), 'Reflecting particle velocity not reflect'
+assert (np.all(vz == +vz0)), 'Periodic particle velocity not reflect'
```
Comment (RemiLehe) [inline Examples/Tests/boundaries/analysis.py]:
```suggestion
assert (np.all(vx == -vx0)), 'Reflecting particle velocity not reflected'
assert (np.all(vz == +vz0)), 'Periodic particle velocity not reflected'
```

File: Source/Particles/ParticleBoundaries.H [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    CheckAll (ParticleBoundaryType bc) {
+        return (xmin_bc == bc && xmax_bc == bc
+#ifdef WARPX_DIM_3D
+             && ymin_bc == bc && ymax_bc == bc
+#endif
+             && zmin_bc == bc && zmax_bc == bc);
+    }
+
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    static void
```
Comment (RemiLehe) [inline Source/Particles/ParticleBoundaries.H:73]: Could you add a docstring to that function, specifying that, depending on the actual boundary conditions, this function:
- marks the particles that need to be removed from the simulation
- Changes the sign of the velocity for reflected particles

**Review by RemiLehe — CHANGES_REQUESTED** [review]
Thanks a lot for this PR! I have a few minor requests, related to code documentation.

File: Docs/source/usage/parameters.rst (hunk 1 above)
Comment (dpgrote) [reply Docs/source/usage/parameters.rst]: This was old documentation that I had forgotten to delete. The correct way now it to set boundary.particle_lo and _hi (and that has been updated).

**Comment by dpgrote** [comment]
Thanks for the review @RemiLehe !

**Comment by RemiLehe** [comment]
Hm, it seems that one of the CI tests crashed here. @dpgrote Would you have time to look into this?

File: Examples/Tests/boundaries/inputs_3d [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+geometry.coord_sys   = 0                  # 0: Cartesian
+geometry.prob_lo     = -1.0  -1.0  -1.0   # physical domain
+geometry.prob_hi     =  1.0   1.0   1.0
+
+boundary.field_lo = pec pec periodic
+boundary.field_hi = pec pec periodic
+boundary.particle_lo = reflecting absorbing periodic
+boundary.particle_hi = reflecting absorbing periodic
+
+warpx.do_pml = 0
```
Comment (RevathiJambunathan) [inline Examples/Tests/boundaries/inputs_3d:25]:
```suggestion
```
This flag is not needed now?

**Review by RevathiJambunathan — COMMENTED** [review]
I just have one minor comment and a question.
What happens to the current deposited? Will the current in the guard cell be flipped after the velocities are flipped?
or nothing specific needs to be done? In my notes on boundary condition for particles, I have a bullet point : "What happens to current? that it should be similar to on-axis RZ and that we can refer to warp -- copy deposited current and mirror it" for reflecting boundary condition

**Comment by dpgrote** [comment]
Hi @RevathiJambunathan - that's a good question about the current in the guard cells. I don't think there is a simple answer. Also, this will depend on the field boundary condition. The mirroring of the current may be needed for both particle absorbing and reflecting boundaries for instance when using a PEC field boundary.

**Comment by RevathiJambunathan** [comment]
Ah I see. Thank you @dpgrote . May be we could meet offline and discuss how to add this in?

**Comment by RemiLehe** [comment]
The PR looks good, but I think that one of the tests is still failing. @dpgrote Would you have time to look into it?

**Comment by RemiLehe** [comment]
It looks like the `particle_boundaries_3d` test is failing in the CI:
```
   mpiexec -n 2 ./main3d.gnu.TEST.TPROF.MTMPI.OMP.QED.OPMD.PSATD.GPUCLOCK.ex inputs_3d  diag1.file_prefix=particle_boundaries_3d_plt   amrex.abort_on_unused_inputs=1
   WARNING: test did not produce any output
   WARNING: no output file.  Skipping visualization
   execution time: 5.0s
   creating problem test report ...
   particle_boundaries_3d CRASHED (backtraces produced)
```

File: Source/Particles/WarpXParticleContainer.H [hunk 6]
```diff
@@ -13,7 +13,7 @@
 #include "WarpXParticleContainer_fwd.H"
 
 #include "Evolve/WarpXDtType.H"
-#include "SpeciesPhysicalProperties.H"
+#include "Particles/ParticleBoundaries.H"
```
Comment (ax3l) [inline Source/Particles/WarpXParticleContainer.H:16]: Accidentally removed in conflict resolution:
```suggestion
#include "Particles/ParticleBoundaries.H"
#include "SpeciesPhysicalProperties.H"
```

File: Source/Particles/ParticleBoundaries.H [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+#include <AMReX_AmrCore.H>
+
+/**
+ *  This allows a different boundary condition on each edge of the grid.
+ */
+struct ParticleBoundaries
+{
+
+    ParticleBoundaries () noexcept
```
Comment (ax3l) [inline Source/Particles/ParticleBoundaries.H:20]: Just saw this: in order to keep compile time for TUs that include this header low, we should probably move all those implementations into a `.cpp` file.

Comment (dpgrote) [reply Source/Particles/ParticleBoundaries.H:20]: I can submit a PR to do this.
